.azure-pipelines/testscripts_analyse/report_data_storage.py

```diff
@@ -0,0 +1,93 @@
+"""Wrappers and utilities for storing test reports."""
+import json
+import os
+import tempfile
+
+from abc import ABC
+from azure.kusto.data import KustoConnectionStringBuilder
+
+try:
+    from azure.kusto.ingest import KustoIngestClient
+except ImportError:
+    from azure.kusto.ingest import QueuedIngestClient as KustoIngestClient
+
+from azure.kusto.ingest import IngestionProperties
+
+# Resolve azure.kusto.ingest compatibility issue
+try:
+    from azure.kusto.ingest import DataFormat
+except ImportError:
+    from azure.kusto.data.data_format import DataFormat
+
+from datetime import datetime
+
+
+class KustoConnector(ABC):
+    """KustoReportDB is a wrapper for storing test reports in Kusto/Azure Data Explorer."""
+
+    TESTSCRIPT_TABLE = "TestScripts"
+
+    TABLE_FORMAT_LOOKUP = {
+        TESTSCRIPT_TABLE: DataFormat.JSON,
+    }
+
+    TABLE_MAPPING_LOOKUP = {
+        TESTSCRIPT_TABLE: "TestScriptsMapping"
+    }
+
+    def __init__(self, db_name: str):
+        """Initialize a Kusto report DB connector.
+
+        Args:
+            db_name: The Kusto database to connect to.
+        """
+        self.db_name = db_name
+
+        """
+            Kusto performance depends on the work load of cluster,
+            to improve the high availability of test result data service
+            by hosting a backup cluster, which is optional.
+        """
+        ingest_cluster = os.getenv("TEST_REPORT_INGEST_KUSTO_CLUSTER_BACKUP")
+        tenant_id = os.getenv("TEST_REPORT_AAD_TENANT_ID_BACKUP")
+        service_id = os.getenv("TEST_REPORT_AAD_CLIENT_ID_BACKUP")
+        service_key = os.getenv("TEST_REPORT_AAD_CLIENT_KEY_BACKUP")
+
+        if not ingest_cluster or not tenant_id or not service_id or not service_key:
+            print("Could not load backup Kusto Credentials from environment")
+            self._ingestion_client_backup = None
```

Comment:
if we cannot connect to kusto, why not throw an error here?

Reply:
Raise a runtime error here.